pkgs/misc/crust/default.nix

```diff
@@ -0,0 +1,154 @@
+###!   $ nix-build -A pkgsCross.or1k.crustTeresA64
+###!
+###! When packaging for linux make sure that you have the following patches otherwise crust won't be able to cleanly share the clock controller and PMIC bus controller hardware: https://github.com/torvalds/linux/compare/master...crust-firmware:linux:crust
+###!
+###! If you are faced with issues while building the firmware then take a look at https://github.com/crust-firmware/crust#building-the-firmware
+
+{ stdenv
+, lib
+, fetchFromGitHub
+, fetchpatch
+, bison
+, flex
+, buildPackages
+}:
+
+###! The package declaration is heavily inspired by the U-Boot package to make maintenance easier, but should stay as an invidual declaration since it's processing has different requirements and including it in the U-Boot would bloat the definition and would be harder to understand and maintain
+
+let
+  defaultVersion = "unstable-2023-03-04";
+  defaultSrc = fetchFromGitHub {
+    owner = "crust-firmware";
+    repo = "crust";
+    rev = "c308a504853e7fdb47169796c9a832796410ece8";
+    sha256 = "AobVD3Jo/6s0cExHXpYYAqZv/gCplxLlDYVscSCIS6M=";
+  };
+  buildCrust = {
+    version ? "${defaultVersion}"
+  , src ? null
+  , filesToInstall
+  , installDir ? "$out"
+  , defconfig
+  , extraConfig ? ""
+  , extraPatches ? []
+  , extraMakeFlags ? []
+  , extraMeta ? {}
+  , ... } @ args: stdenv.mkDerivation ({
+    # Strip "_defconfig" suffix from 'defconfig' if it's present for clearer output
+    # builtins.match returns 'Null' if the string matches so we have to then check if it's Null
+    # FIXME-QA(Krey): I hate this, but am not aware of better way to write this
+    pname = if (builtins.isNull ((builtins.match "_defconfig$" defconfig)))
+      then "crust-${lib.strings.removeSuffix "_defconfig" defconfig}"
+      else "crust-${defconfig}";
+
+    ###! The package has upstream definitions of configurations for set devices and thus the defconfig has to be always included, beyond that the options are designed to be as flexible as possible to accompany special requirements of various devices
+    ###!
+    ###! Avoid declaring extra configuration (extraConfig) when possible as those should go to the upstream unless you use special usecase device e.g. modded teres-A64
+
+    defconfig = if defconfig == null then
+      throw "crust config not defined!"
+      else defconfig;
+
+    version = if src == null then
+      defaultVersion
+      else version;
+
+    src = if src == null then
+      defaultSrc
+      else src;
+
+    # Only for global crust patches, per-device paches should be included as extras to avoid adding noise which would complicate debugging
+    patches = [] ++ extraPatches;
+
+    postPatch = "";
+
+    nativeBuildInputs = [
+      flex
+      bison
+    ];
+
+    depsBuildBuild = [ buildPackages.stdenv.cc ];
+
+    makeFlags = [
+      # This is double-quotted to prevent string splitting (SC2086)
+      "CROSS_COMPILE=\"${stdenv.cc.targetPrefix}\""
+    ] ++ extraMakeFlags;
+
```

Comment:
Multiple problems:

 - We won't need `extraMakeFlags`, and if it ever happens (unlikely, as it is a simple project) we'll just add it then.
 - We don't need to do quoting here, as it's already being handled intrinsically by it being part of the `makeFlags` as an array. `shellcheck` couldn't see it as otherwise.

In fact, I'm kinda surprised it works, it must be a `make` misfeature regarding quoting that makes it work. Tracing with `set-x`:

```
++ make [...] 'CROSS_COMPILE="or1k-elf-"'
```

It's been double-escaped.

Reply:
> We won't need extraMakeFlags, and if it ever happens (unlikely, as it is a simple project) we'll just add it then. -- @samueldr (https://github.com/NixOS/nixpkgs/pull/241360/files#r1253843921)

Disagree with the approach considering usecase and scope of the device. I used the `extraMakeFlags` when testing hardening during the development of the package and plan on making userland changes that utilizes this declaration for drone development.

> We don't need to do quoting here, as it's already being handled intrinsically by it being part of the makeFlags as an array. shellcheck couldn't see it as otherwise. -- @samueldr (https://github.com/NixOS/nixpkgs/pull/241360/files#r1253843921)


###### Scenario A:
```nix
makeFlags = [
  "CROSS_COMPILE=\"${stdenv.cc.targetPrefix}\" KREYREN"
] ++ extraMakeFlags;
```

>  > build flags: SHELL=/nix/store/51sszqz1d9kpx480scb1vllc00kxlx79-bash-5.2-p15/bin/bash CROSS_COMPILE=\"or1k-elf-\" KREYREN
>  > make: *** No rule to make target 'KREYREN'.  Stop.

###### Scenario B:

```nix
makeFlags = [
  "CROSS_COMPILE=\"${stdenv.cc.targetPrefix} KREYREN\""
] ++ extraMakeFlags;
```

>  > build flags: SHELL=/nix/store/51sszqz1d9kpx480scb1vllc00kxlx79-bash-5.2-p15/bin/bash CROSS_COMPILE=\"or1k-elf- KREYREN\"
>  > /nix/store/51sszqz1d9kpx480scb1vllc00kxlx79-bash-5.2-p15/bin/bash: -c: line 1: unexpected EOF while looking for matching `"'
>  > /nix/store/51sszqz1d9kpx480scb1vllc00kxlx79-bash-5.2-p15/bin/bash: -c: line 1: unexpected EOF while looking for matching `"'
>  > /nix/store/51sszqz1d9kpx480scb1vllc00kxlx79-bash-5.2-p15/bin/bash: -c: line 1: unexpected EOF while looking for matching `"'
>  > make: *** No rule to make target 'KREYREN"'.  Stop.

---

The attempt here was that in case `${stdenv.cc.targetPrefix}` is malformed that it would fail to build under make through parsing invalid argument in the builder command instead of parsing arguments in make as those have higher likelyhood of causing unwanted outcome and may cause make to continue building without good output:

```nix
makeFlags = [
  "CROSS_COMPILE=${stdenv.cc.targetPrefix} LD="
] ++ extraMakeFlags;
```

>  > collect2: error: ld returned 1 exit status
>  > make: *** [Makefile:201: build/scp/scp.elf] Error 1

^ Arguments parsed in LD

But the double-quotes are escaped and i am not aware of a way to unescape them within MakeFlags so i will use your suggestion.

> In fact, I'm kinda surprised it works, it must be a make misfeature regarding quoting that makes it work. Tracing with set-x: -- @samueldr (https://github.com/NixOS/nixpkgs/pull/241360/files#r1253843921)

I was told that this is by design so that make works in various edge cases.